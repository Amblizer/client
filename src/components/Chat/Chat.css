/* MESSANGER */

.chat-innerWrapper-messenger {
  display: block;
  width: 280px;
  height: 360px;
  box-shadow: #dddddd 1px 1px 8px 0;
  border-width: 0 1px 1px 0;
  border-radius: 3px 3px 0 0;
  border-style: solid;
  border-color: #cccccc;
  border-bottom: 0;
}

.chat-outerWrapper-messenger {
  position: relative;
  width: 80%;
  margin: 0 auto;
}

.chat-box-messenger {
  position: absolute;
  right: 0;
  bottom: 0;
  background-color: #ffffff;
}

.chat-header-messenger {
  display: block;
  width: 100%;
  height: 32px;
  padding: 8px;
  box-sizing: border-box;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.15);
  border-radius: 3px 3px 0 0;
  font-size: 13px;
  font-family: 'Arial', sans-serif;
  color: #ffffff;
  background: #72a0e5;
}

.chat-headerText-messenger {}

.chat-icon-messenger {
  border: 0;
  float: right;
  font-size: 24px;
  margin-right: -8px;
  margin-top: -7px;
  font-weight: 800;
  cursor: pointer;
  color: rgba(255,255,255,0.5);
  background: transparent;
}


/* FLOAT */

.chat-innerWrapper-float {
  display: block;
  width: 280px;
  height: 360px;
}

.chat-outerWrapper-float {
  position: relative;
  width: 90%;
  margin: 0 auto;
}

.chat-box-float {
  position: absolute;
  right: 0;
  bottom: 120px;
  background-color: transparent;
}

.chat-header-float {
  position: absolute;
  bottom: -48px;
  right: -48px;
  display: block;
  width: 42px;
  height: 42px;
  padding: 11px 5px 11px 6px;
  box-sizing: border-box;
  box-shadow: #dddddd 1px 1px 8px 0;
  border-radius: 21px;
  font-size: 13px;
  font-family: 'Arial', sans-serif;
  color: #ffffff;
  background: #72a0e5;
}

.chat-headerText-float {
  display: none;
}

.chat-icon-float {
  float: right;
  font-size: 24px;
  font-weight: 800;
  margin-top: -5px;
  border: 0;
  cursor: pointer;
  color: rgba(255,255,255,0.5);
  background: transparent;
}

/* SIDEPANEL */

.chat-innerWrapper-sidepanel {
  display: block;
  width: 280px;
  height: 100%;
  box-shadow: #dddddd 1px 1px 8px 0;
  border-width: 0 0 1px 0;
  border-radius: 3px 3px 0 0;
  border-style: solid;
  border-color: #cccccc;
  border-bottom: 0;
}

.chat-outerWrapper-sidepanel {
  position: relative;
  height: 100%;
  width: 80%;
  margin: 0 auto;
}

.chat-box-sidepanel {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  background-color: #ffffff;
}

.chat-header-sidepanel {
  display: none;
}

.chat-headerText-sidepanel {}

.chat-icon-sidepanel {
  border: 0;
  float: right;
  font-size: 24px;
  margin-right: -10px;
  margin-top: -7px;
  font-weight: 800;
  cursor: pointer;
  color: rgba(255,255,255,0.5);
  background: transparent;
}