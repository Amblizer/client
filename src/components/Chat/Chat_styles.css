.Chat-outerWrapper {
  position: relative;
  width: 80%;
  margin: 0 auto;
}

.Chat-box {
  position: absolute;
  right: 0;
  bottom: 0;
  background-color: #ffffff;
}

.Chat-innerWrapper {
  display: block;
  width: 280px;
  height: 360px;
  box-shadow: #dddddd 1px 1px 8px 0;
  border-width: 0 1px 1px 0;
  border-radius: 3px 3px 0 0;
  border-style: solid;
  border-color: #cccccc;
  border-bottom: 0;
}

.Chat-header {
  display: block;
  width: 100%;
  height: 32px;
  padding: 8px;
  box-sizing: border-box;
  border-radius: 3px 3px 0 0;
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.15);
  font-size: 13px;
  font-family: \Arial\;
  sans-serif;
  color: white;
  background: #ef7f7f;
}

.Chat-headerText {}

.Chat-icon {
  border: 0;
  float: right;
  font-size: 24px;
  margin-right: -8px;
  margin-top: -7px;
  font-weight: 800;
  cursor: pointer;
  color: rgba(255, 255,255, 0.5);
  background: rgba(0, 0, 0, 0);
}
