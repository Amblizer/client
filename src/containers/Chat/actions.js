import {
  CHAT_MESSAGE_CLIENT,
  CHAT_MESSAGE_OPERATOR,
} from './constants';

export function sendMessage(payload) {
  return {
    type: CHAT_MESSAGE_CLIENT,
    payload,
  };
}

export function recieveMessage(payload) {
  return {
    type: CHAT_MESSAGE_OPERATOR,
    payload,
  };
}
