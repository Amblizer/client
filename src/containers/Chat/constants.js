export const CHAT_CONNECTED = 'CHAT_CONNECTED';
export const CHAT_DISCONNECTED = 'CHAT_DISCONNECTED';
export const CHAT_MESSAGE_CLIENT = 'CHAT_MESSAGE_CLIENT'; // Check
export const CHAT_MESSAGE_OPERATOR = 'CHAT_MESSAGE_OPERATOR'; // Check

export const CHAT_CLIENT = 'CLIENT';
export const CHAT_OPERATOR = 'OPERATOR';
