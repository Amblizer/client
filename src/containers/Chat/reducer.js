import {
  CHAT_CONNECTED,
  CHAT_DISCONNECTED,
  CHAT_MESSAGE_CLIENT,
  CHAT_MESSAGE_OPERATOR,
  CHAT_CLIENT,
  CHAT_OPERATOR,
} from './constants.js'


// Chat based actions (send, recieve) will go through the chatReducer
  const chatInitialState = {
    // messages: [
    //   {author: 'CLIENT', content: ['Im a client']},
    //   {author: 'OPERATOR', content: ['Im an operator']}
    //   ],
    messages:[]
  };
  const chatReducer = function ChatReducer (state = chatInitialState, action) {
    let messages = [];

    console.log('CHAT', action.type);

    switch (action.type) {
      case CHAT_MESSAGE_OPERATOR:
        messages = state.messages;

        // Is the last message from client? (e.g. can we combine it)
        if (messages.length > 0 && messages[messages.length - 1].author === CHAT_OPERATOR) {
          messages[messages.length - 1].content.push(action.payload);
        } else {
          messages.push({
            author: CHAT_OPERATOR,
            content: [action.message],
          });
        }

        return Object.assign({}, state, {
          messages,
        });

      case CHAT_MESSAGE_CLIENT:
        console.log('chat message from client, content is', action.payload)
        messages = state.messages;

        // Is the last message from client? (e.g. can we combine it)
        if (messages.length > 0 && messages[messages.length - 1].author === CHAT_CLIENT) {
          messages[messages.length - 1].content.push(action.payload);
        } else {
          messages.push({
            author: CHAT_CLIENT,
            content: [action.payload],
          });
        }
        console.log(messages);

        return Object.assign({}, state, {
          messages,
        });
        
      default:
        return state;
    }
  };

  export default chatReducer;