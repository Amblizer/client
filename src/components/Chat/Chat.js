import React, {PropTypes} from 'react';
import { connect } from 'react-redux';
import io from 'socket.io-client';
import './Chat_styles.css';
import Messages from '../Messages/Messages';
import Input from '../Input/Input';
import {rotateChatStyle} from '../../containers/UI/actions.js'

const socketPath = 'http://localhost:8000';

// Create chat box wrapper
class Chat extends React.Component {
  static propTypes = {
  }

  constructor (props) {
    super(props);

    this.socket = io.connect(socketPath, {
      reconnectionAttempts: 10,
    });

    this.socket.on('connect', this.onSocketConnected);
    this.socket.on('connect_error', this.onSocketConnectionError);
    this.socket.on('connect_timeout', this.onSocketTimeout);
    this.socket.on('disconnect', this.onSocketDisconnected);
    this.socket.on('reconnect', this.onSocketReconnected);
    this.socket.on('reconnecting', this.onSocketReconnecting);
    // this.socket.on('reconnect_error', socketConnectionError);
    this.socket.on('reconnect_failed', this.onSocketReconnectionFailed);
    this.socket.on('reconnect_timeout', this.onSocketTimeout);

    this.socket.on('operator:message', this.handleOperatorMessage);

    // Initial state
    this.state = {
      operator: {
        firstName: 'John',
      },
      company: {
        name: 'ACME',
      },
    };
  }


  // Event Handlers

  // Successful connection
  onSocketConnected() {
    // const state = store.getState();
    console.log('DEBUG', 'Socket connected');

    // if (!this.props.state.chat.connected) {
    //   store.dispatch({ type: CHAT_CONNECTED });
    // }
  }

  // Disconnected
  onSocketDisconnected() {
    const state = store.getState();
    console.warn('DEBUG', 'Socket disconnected');

    if (state.chat.connected) {
      store.dispatch({ type: CHAT_DISCONNECTED });
    }
  }

  // Successful re-connected
  onSocketReconnected () {
    console.log('DEBUG', 'Socket reconnected');
  }

  // Attempting to re-connect
  onSocketReconnecting () {
    console.log('DEBUG', 'Socket reconnecting ...');
  }

  // Failed to re-connect after manager.reconnectionAttempts tried
  onSocketReconnectionFailed () {
    console.error('DEBUG', 'Socket failed reconnection');
  }

  // Timeout either connecting or re-connecting
  onSocketTimeout () {
    console.warn('DEBUG', 'Socket timeout');
  }

  // Error when connecting or re-connecting
  onSocketConnectionError () {
    console.error('DEBUG', 'Socket connection error');
  }

  handleOperatorMessage (data) {
    console.log('DEBUG', 'RECIEVING MESSAGE ...', data);

    store.dispatch({ type: CHAT_MESSAGE_OPERATOR, message: data });
  }


  // Actions

  open () {
    console.log('DEBUG', 'Open chat');
  }

  close () {
    console.log('DEBUG', 'Close chat');
  }


  render () {
    // const { sheet: { classes } } = this.props;
    const socket = this.socket;
    const operator = this.state.operator;
    const company = this.state.company;
    const {chatStyle, dispatch} = this.props 
    

    return (
      <div className={`Chat-outerWrapper_${chatStyle}`}>
        <div className={`Chat-box_${chatStyle}`}>
          <div className={`Chat-innerWrapper_${chatStyle}`}>
            <div className={`Chat-header_${chatStyle}`}>
              <span className={`Chat-headerText_${chatStyle}`}>
                <strong>{operator.firstName}</strong>
                &nbsp;from&nbsp;{company.name}

              </span>

              <div className="demo-rotate"> {/*some demo buttons for testing redux / changing ui state*/}
                <button onClick={() => dispatch(rotateChatStyle('FLOAT'))}> FLOAT </button>
                <button onClick={() => dispatch(rotateChatStyle('MESSENGER'))}> MESSENGER </button>
                <button onClick={() => dispatch(rotateChatStyle('SIDEPANEL'))}> SIDEPANEL </button>
              </div>

              <button className={`Chat-icon_${chatStyle}`} onClick={this.close}>&#215;</button>
            </div>
            <Messages socket={socket} />
            <Input socket={socket} />
          </div>
        </div>
      </div>
    );
  }
}

export default connect(
	state => ({chatStyle: state.ui.chatStyle}),
	dispatch => ({dispatch})
	)(Chat);
