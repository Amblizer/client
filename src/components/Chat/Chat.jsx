import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

import io from 'socket.io-client';

import Messages from '../Messages/Messages.jsx';
import Input from '../Input/Input.jsx';
import './Chat.css';

import { openChat, toggleChat, closeChat } from '../../containers/UI/actions';
import { recieveMessage } from '../../containers/Chat/actions';

const socketPath = 'http://localhost:8000';

// Create chat box wrapper
export class ChatComponent extends Component {
  static propTypes = {
    chatStyle: PropTypes.string,
    dispatch: PropTypes.func,
    open: PropTypes.bool,
  }

  constructor (props) {
    super(props);

    this.socket = io.connect(socketPath, {
      reconnectionAttempts: 10,
    });

    this.socket.on('connect', this.onSocketConnected);
    this.socket.on('connect_error', this.onSocketConnectionError);
    this.socket.on('connect_timeout', this.onSocketTimeout);
    this.socket.on('disconnect', this.onSocketDisconnected);
    this.socket.on('reconnect', this.onSocketReconnected);
    this.socket.on('reconnecting', this.onSocketReconnecting);
    // this.socket.on('reconnect_error', socketConnectionError);
    this.socket.on('reconnect_failed', this.onSocketReconnectionFailed);
    this.socket.on('reconnect_timeout', this.onSocketTimeout);

    this.socket.on('operator:message', this.handleOperatorMessage());

    // Initial state
    this.state = {
      operator: {
        firstName: 'John',
      },
      company: {
        name: 'ACME',
      },
    };
  }


  // Event Handlers

  // Successful connection
  onSocketConnected() {
    // const state = store.getState();
    console.log('DEBUG', 'Socket connected');

    // if (!this.props.state.chat.connected) {
    //   store.dispatch({ type: CHAT_CONNECTED });
    // }
  }

  // Disconnected
  onSocketDisconnected() {
    // const state = store.getState();
    console.warn('DEBUG', 'Socket disconnected');

    // if (state.chat.connected) {
    //   store.dispatch({ type: CHAT_DISCONNECTED });
    // }
  }

  // Successful re-connected
  onSocketReconnected () {
    console.log('DEBUG', 'Socket reconnected');
  }

  // Attempting to re-connect
  onSocketReconnecting () {
    console.log('DEBUG', 'Socket reconnecting ...');
  }

  // Failed to re-connect after manager.reconnectionAttempts tried
  onSocketReconnectionFailed () {
    console.error('DEBUG', 'Socket failed reconnection');
  }

  // Timeout either connecting or re-connecting
  onSocketTimeout () {
    console.warn('DEBUG', 'Socket timeout');
  }

  // Error when connecting or re-connecting
  onSocketConnectionError () {
    console.error('DEBUG', 'Socket connection error');
  }

  handleOperatorMessage () {
    const { dispatch } = this.props;

    return function handleOperatorMessageCurry (data) {
      console.log('DEBUG', 'RECIEVING MESSAGE ...', data);

      dispatch(recieveMessage(data));
    };
  }


  // Actions

  open () {
    const { dispatch } = this.props;

    console.log('DEBUG', 'Open chat');

    dispatch(openChat());
  }

  toggle () {
    const { dispatch } = this.props;

    console.log('DEBUG', 'Toggling chat');

    dispatch(toggleChat());
  }

  close () {
    const { dispatch } = this.props;

    console.log('DEBUG', 'Close chat');

    dispatch(closeChat());
  }

  // This is where the magic happens

  render () {
    // const { sheet: { classes } } = this.props;
    const socket = this.socket;
    const operator = this.state.operator;
    const company = this.state.company;
    const { open, chatStyle } = this.props;

    const chatStyleName = chatStyle.toLowerCase();
    const chatClasses = [
      `chat-outerWrapper-${chatStyleName}`,
    ];

    if (!open) {
      chatClasses.push('closed');
    }

    return (
      <div className={chatClasses.join(' ')}>
        <div className={`chat-box-${chatStyleName}`}>
          <div className={`chat-innerWrapper-${chatStyleName}`}>
            <div className={`chat-header-${chatStyleName}`}>
              <span className={`chat-headerText-${chatStyleName}`}>
                <strong>{operator.firstName}</strong>&nbsp;from&nbsp;{company.name}
              </span>
              <button className={`chat-icon-${chatStyleName}`} onClick={this.toggle}>&#215;</button>
            </div>
            <Messages socket={socket} />
            <Input socket={socket} />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  chatStyle: state.ui.chatStyle,
  open: state.ui.open,
});

const mapDispatchToProps = dispatch => ({
  dispatch,
});

const Chat = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ChatComponent);

export default Chat;
