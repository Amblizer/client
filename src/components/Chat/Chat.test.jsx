import React from 'react';
import { Provider } from 'react-redux';
import io from 'socket.io-client';

import { shallow, render } from 'enzyme';

import Chat from './Chat';

const socket = {
  on: jest.fn(),
};

const store = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
  getState: jest.fn(() => ({
    ui: { chatStyle: 'TEST' },
    chat: { messages: [] },
  })),
};

io.connect = jest.fn(() => socket);

describe('Chat', () => {
  it('matches snapshot', () => {
    const component = shallow(<Chat store={store} />);

    expect(component).toMatchSnapshot();
  });

  it('starts a socket connection', () => {
    // Requires a store/Provider for render or mount enzyme functions for
    //  some reason
    render(
      <Provider store={store}>
        <Chat />
      </Provider>,
    );

    expect(io.connect).toHaveBeenCalled();
  });
});
