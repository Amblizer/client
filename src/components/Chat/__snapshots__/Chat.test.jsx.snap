exports[`Chat matches snapshot 1`] = `
<ChatComponent
  chatStatus="TEST"
  chatStyle="TEST"
  dispatch={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  } />
`;
